doc/developer/design/20210811_catalog_only_mode.md

```diff
@@ -0,0 +1,115 @@
+# Catalog-only mode
+
+## Summary
+
+When users have views causing OOM crashes, we want to provide a mechanism that
+still lets users boot into Materialize and take actions to remediate the issue,
+e.g. drop the OOMing view. To accomplish this, we'll introduce a new boot mode
+called "catalog-only mode" that starts Materialize as normal with the exception
+of not shipping any dataflows for user-created objects, such as materialized
+views.
+
+Ideally, we'll also provide a mechanism that lets users enable dataflows for
+their objects while Materialize is running using a statement such as `ALTER
+INDEX <index name> SET ENABLED`.
+
+## Goals
+
+- Support booting in catalog-only mode, which will not:
+    - Ship the dataflow for any user-created object
+    - Support querying user-created objects, such as views
+    - Support inserting into tables
+
+    Any other action should be allowed, such as creating, modifying, or dropping
+    objects.
+- Introduce a `--catalog-only` flag to Materialize that lets users easily turn
+  on catalog-only mode.
+- Add support for `ALTER INDEX ... SET ENABLED` to enable dataflows for objects
+  while Materialilze is running, and lift the restrictions imposed on
+  catalog-only mode for objects that use those dataflows.
+
+  Note that this statement will be "intrinsically idempotent," so `ALTER INDEX
+  ... SET ENABLED` will always succeed, even if Materialize was not booted in
+  catalog only mode or its dataflows are already active. This matches PG's
+  behavior for `ALTER` statements like `ALTER TABLE a ALTER COLUMN x SET NOT
+  NULL`, which you can run repeatedly.
+- Add no more than a single boolean check to the code path of any operation
+  occurring outside of catalog-only mode.
+
+## Non-Goals
+
+- We are not planning to add a statement to re-enable the dataflows for all
+  user-created objects; users can either accomplish this with a script or by
+  simply rebooting _not_ in catalog-only mode.
+
+## Description
+
+The `Coordinator` will handle the enforcing catalog-only mode on objects, i.e.
```

Comment:
How will sinks work with this?

Reply:
That's a good question! I'm not totally sure, but my fuzzy intuition around it was that the only sinks that could still receive data would be those based on system views; everything else in the system exists in a form of stasis. Once you start resuming the activity on your indexes, though, I believe data could start getting fed into sinks.

Reply:
I think probably you'd want to require `ALTER SINK ... SET ENABLED` too. Sinks and indexes are sort of similar in that way, in that they both cause work to happen.